### ISSUE TITLE
Training data resolution?

### ISSUE THREAD

**Issue by markemus** [AdalbertoCq/Pathology-GAN#7]
I'm trying to build my own training set for the model. What was the magnification ratio for the training dataset (40x, 20x, 10x)? I wasn't able to find this in the paper or documentation.

**Comment by AdalbertoCq** [comment]
Hi markemus,

The images are from the [Stanford TMA database](https://tma.im/cgi-bin/home.pl). I wasn't able to find the the magnification ratio, neither in the [Beck et al](https://stm.sciencemag.org/content/3/108/108ra113/tab-figures-data) paper that first used them.

I have used the model on other types of tissue and datasets, it tends to work better with higher magnification ratios (20x and higher).  

Hope this helps. Thanks,
Adal

**Comment by markemus** [comment]
Thanks Adalberto. I've been training for the past couple of days on one of our own datasets (40x, ~600k tiles) and the generation comes out looking swirly and bubbly. Is this a failure mode you've seen before? Does it just need more training? This is after 21 epochs.

https://imgur.com/a/k6oTOj9

I trained before this on your own dataset and the generation looked reasonable even after just 1 epoch.

**Comment by AdalbertoCq** [comment]
Do you have some examples of the real images too? 

The number of tiles seems reasonable, the GAN should be able to handle them. It should also show some convergence after 1 epoch. Looking at the image on epoch 21, I would guess that the GAN keeps collapsing, are you seeing a noisy loss the generator? I would try to increase the gradient penalty on the discriminator, gp_coef (inside the run_pathgan.py) to 0.8-1.0. 

Something else that I have seen is that the model has an easier time to converge when you have less amount tiles with background in the images. Having a stricter threshold on discarding tiles with a % of background usually helps to get better image quality. 

I would start first with a large % of tissue in the tile (like 90% for example) and then tune it down to whatever fits best for you.

Thanks,
Adal

**Comment by markemus** [comment]
Hey Adal,

Here are three examples from the training data. I pulled them from the training h5 file so I could double check it was formatted properly. The tiles comes from pathologist annotated ROIs, so there should be very little background in the dataset.

https://imgur.com/a/BMi3Xab

I'm restarting training from scratch in case it fixes itself, if not I'll start experimenting with the hyperparams you mentioned.

What's the normal range of loss you expect for the generator?

**Comment by AdalbertoCq** [comment]
Sounds good, those images look good. 

It's definitely a higher magnification that I've worked with before but doesn't seem too different so it won't work. The little background sounds good too.

As a general guideline, for the first two epochs it's alright if the generator/discriminator losses are a bit noisy. After that the generator should get to values between 4-2 and stabilize, for the discriminator between 1-0.5. 

I think changing the gp_coeff should make a difference in stabilizing the training of the generator. 

If you have more questions, don't hesitate to send me an email if you want too. 

Thanks,
Adal

**Comment by markemus** [comment]
Hey @AdalbertoCq , just wanted to follow up- unfortunately we weren't able to get it working for our 40x data. We tried adjusting the gp_coeff iteratively as high as 1.2, but we kept having the same issue so we gave up. Maybe the next dataset! Thanks for your help and it really was a disappointment- this is still far and away the most realistic data generator we've seen.

**Comment by AdalbertoCq** [comment]
Hey @markemus,

Sorry to hear that. I'm happy to give a hand and check if I can make it work if you guys want/can share the data. 

We have been able to run the model with different cancer tissue datasets, from TCGA WSIs to local datasets. From the information you gave, it seems reasonable to assume it will work. 

If you want to share some of the runs, I can check those too. 

Thanks,
Adal
